File: Algorithms/134. Gas Station/Solution.java
public class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
		int[] n = new int[gas.length];
		for (int i = 0; i < n.length; i++) {
			n[i] = gas[i] - cost[i];
		}
		int sum = 0;
		int begin = 0;
		int truebegin=0;
		for (int i = 0; i < n.length*2; i++) {
			sum += n[i%n.length];
			while (sum < 0) {
				sum -= n[truebegin];
				begin++;
				truebegin=begin%n.length;
			}
			if (i-begin>=n.length) {
				return truebegin;
			}
		}
		return -1;
	}
}

File: Algorithms/134. Gas Station/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        l=len(gas)
        remider=[gas[i]-cost[i] for i in range(l)]
        if sum(remider)<0:
            return -1
        ans=0
        flag=True
        while ans<l and flag:
            flag=False
            total=0
            for i in range(ans,ans+l):
                total+=remider[i%l]
                if total<0:
                    if ans>i%l+1:
                        return -1
                    ans=i%l+1
                    flag=True
                    break
        return ans%l